src/ZEO/asyncio/client.py

```diff
@@ -620,10 +626,7 @@ def load_before_threadsafe(self, future, wait_ready, oid, tid):
     @future_generator
     def _prefetch(self, oid, tid):
         try:
-            data = yield self.protocol.load_before(oid, tid)
-            if data:
-                data, start, end = data
-                self.cache.store(oid, start, end, data)
+            yield self.protocol.load_before(oid, tid)
```

Comment:
Oh I missed that, thanks.

Is is tested ? I'm not asking about the race, but maybe not doing this change would break prefetch. Or would it just become dead code (i.e. data always false) ?

Reply:
You are welcome, and yes - prefetech is tested:

https://github.com/zopefoundation/ZEO/blob/3d90ed421b22b403ae32cf5d76ced53409479ecb/src/ZEO/tests/testZEO.py#L1071-L1113

What is removed is just a duplicate of cache update that is now living in `protocol.load_before`. It seems to be it would not be a bug to leave cache.store here because it would only store into the cache what was just stored into the cache by protocol.load_before itself. The wakup here seems to be immediate after switching load_before from `asyncio.Future` to `Fu`t, so there is seemingly no race window with invalidations. At least I could not make a data corruption with reintroducing this block back and injecting prefetches into my zloadrace.

Anyway, since prefech is just semantically "call loadBefore, but do not wait for it to complete", and `load_before_threadsafe` - that is serving `loadBefore` now is not doing cache update and delegates cache update to `protocol.load_before`, this change should be correct.